apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,872 @@
+        // However, we cap it to collateral amount to double check
+        uint256 unchallengedAmount = settlementOffer >= collateralAmount ? collateralAmount : (collateralAmount - settlementOffer);
+        uint256 slashedAmount = collateralAmount - unchallengedAmount;
+
+        challenge.state = ChallengeState.Settled;
+        _unchallengeBalance(submitter, unchallengedAmount);
+        _slashBalance(submitter, challenger, slashedAmount);
+        _returnArbitratorFees(challenge);
+        emit ActionSettled(_actionId);
+    }
+
+    function disputeChallenge(uint256 _actionId) external {
+        (Action storage action, Setting storage setting) = _getActionAndSetting(_actionId);
+        require(_canDispute(action, setting), ERROR_CANNOT_DISPUTE_ACTION);
+        require(msg.sender == action.submitter, ERROR_SENDER_NOT_ALLOWED);
+
+        Challenge storage challenge = action.challenge;
+        uint256 disputeId = _createDispute(action, setting);
+        challenge.state = ChallengeState.Disputed;
+        challenge.disputeId = disputeId;
+        disputes[disputeId].actionId = _actionId;
+        emit ActionDisputed(_actionId);
+    }
+
+    function submitEvidence(uint256 _disputeId, bytes _evidence, bool _finished) external {
+        (Action storage action, Dispute storage dispute) = _getActionAndDispute(_disputeId);
+        require(_canSubmitEvidence(action), ERROR_CANNOT_SUBMIT_EVIDENCE);
+
+        bool finished = _registerEvidence(action, dispute, msg.sender, _finished);
+        _submitEvidence(_disputeId, msg.sender, _evidence, _finished);
+        if (finished) {
+            Setting storage setting = _getSetting(action);
+            setting.arbitrator.closeEvidencePeriod(_disputeId);
+        }
+    }
+
+    function executeRuling(uint256 _actionId) external {
+        (Action storage action, Setting storage setting) = _getActionAndSetting(_actionId);
+        require(_canRuleDispute(action), ERROR_CANNOT_RULE_ACTION);
+
+        uint256 disputeId = action.challenge.disputeId;
+        setting.arbitrator.executeRuling(disputeId);
+    }
+
+    function rule(uint256 _disputeId, uint256 _ruling) external {
+        (Action storage action, Dispute storage dispute) = _getActionAndDispute(_disputeId);
+        require(_canRuleDispute(action), ERROR_CANNOT_RULE_ACTION);
+
+        Setting storage setting = _getSetting(action);
+        IArbitrator arbitrator = setting.arbitrator;
+        require(msg.sender == address(arbitrator), ERROR_SENDER_NOT_ALLOWED);
+
+        dispute.ruling = _ruling;
+        emit Ruled(arbitrator, _disputeId, _ruling);
+
+        if (_ruling == DISPUTES_RULING_SUBMITTER) {
+            _rejectChallenge(action, setting);
+            emit ActionAccepted(dispute.actionId);
+        } else if (_ruling == DISPUTES_RULING_CHALLENGER) {
+            _acceptChallenge(action, setting);
+            emit ActionRejected(dispute.actionId);
+        } else {
+            _voidChallenge(action, setting);
```

Comment:
In this case, is it possible that Aragon Court could have refunded fees (in case no jurors voted for example) and do we want to handle it?

Reply:
We shouldn't handle that here cause it will occur when penalties are settled in court. But it's an edge case we should consider somehow.